File: app/src/main/java/com/example/solar_alarm/AlarmList/AlarmListFragment.java
package com.example.solar_alarm.AlarmList;

import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.solar_alarm.CreateAlarm.UpdateAlarmFragment;
import com.example.solar_alarm.Data.Alarm;
import com.example.solar_alarm.R;
import com.example.solar_alarm.Service.GpsTracker;

import java.time.ZoneId;
import java.util.List;
import java.util.TimeZone;

public class AlarmListFragment extends Fragment implements OnToggleAlarmListener {
    private AlarmRecycleViewAdapter alarmRecyclerViewAdapter;
    private AlarmListViewModel alarmsListViewModel;
    private RecyclerView alarmsRecyclerView;
    private Button addAlarm;
    private Button addLocation;
    private GpsTracker gpsTracker;
    TextView timeZone;
    TextView latitude;
    TextView longitude;
    ZoneId zoneId;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        alarmRecyclerViewAdapter = new AlarmRecycleViewAdapter(this);
        alarmsListViewModel = ViewModelProviders.of(this).get(AlarmListViewModel.class);
        alarmsListViewModel.getAlarmsLiveData().observe(this, new Observer<List<Alarm>>() {
            @Override
            public void onChanged(List<Alarm> alarms) {
                if (alarms != null) {
                    alarmRecyclerViewAdapter.setAlarms(alarms);
                }
            }
        });
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_listalarms, container, false);

        alarmsRecyclerView = view.findViewById(R.id.fragment_listalarms_recylerView);
        alarmsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        alarmsRecyclerView.setAdapter(alarmRecyclerViewAdapter);
        this.configureOnClickRecyclerView();

        zoneId = TimeZone.getDefault().toZoneId();
        timeZone = view.findViewById(R.id.fragment_listalarms_timezone);
        latitude = view.findViewById(R.id.fragment_listalarms_latitude);
        longitude = view.findViewById(R.id.fragment_listalarms_longitude);

        timeZone.setText(zoneId.toString());
        addAlarm = view.findViewById(R.id.fragment_listalarms_addAlarm);
        addAlarm.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Navigation.findNavController(v).navigate(R.id.action_alarmsListFragment_to_createAlarmFragment);
            }
        });
        addLocation = view.findViewById(R.id.fragment_listAlarms_addLocation);
        addLocation.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Navigation.findNavController(v).navigate(R.id.action_alarmsListFragment_to_addLocationFragment);
            }
        });

        getLocation(view);

        return view;
    }

    @Override
    public void onToggle(Alarm alarm) {
        if (alarm.isStarted()) {
            alarm.cancelAlarm(getContext());
            alarmsListViewModel.update(alarm);
        } else {
            alarm.schedule(getContext());
            alarmsListViewModel.update(alarm);
        }
    }

    public void getLocation(View view){
        gpsTracker = new GpsTracker(view.getContext());
        if(gpsTracker.canGetLocation()){
            double lat = gpsTracker.getLatitude();
            double lon = gpsTracker.getLongitude();
            latitude.setText(String.valueOf(lat));
            longitude.setText(String.valueOf(lon));
        }else{
            gpsTracker.showSettingsAlert();
        }
    }

    private void configureOnClickRecyclerView(){
        ItemClickSupport.addTo(alarmsRecyclerView, R.layout.item_alarm)
                .setOnItemClickListener(new ItemClickSupport.OnItemClickListener() {
                    @Override
                    public void onItemClicked(RecyclerView recyclerView, int position, View v) {
                        Alarm alarm = alarmRecyclerViewAdapter.getAlarm(position);
                        Bundle bundle = new Bundle();
                        bundle.putInt("position", position);
                        UpdateAlarmFragment updateAlarmFragment = new UpdateAlarmFragment();
                        updateAlarmFragment.setArguments(bundle);
                        FragmentManager manager = getFragmentManager();
                        manager.beginTransaction().replace(R.id.activity_main_nav_host_fragment, updateAlarmFragment).commit();

                    }
                });
        ItemClickSupport.addTo(alarmsRecyclerView, R.layout.item_alarm)
                .setOnItemLongClickListener(new ItemClickSupport.OnItemLongClickListener() {
                    @Override
                    public boolean onItemLongClicked(RecyclerView recyclerView, int position, View v) {
                        Alarm alarm = alarmRecyclerViewAdapter.getAlarm(position);
                        // 2 - Show result in a Toast
                        Toast.makeText(getContext(), "You long clicked on user : "+alarm.getTitle(), Toast.LENGTH_SHORT).show();
                        alarmsListViewModel.delete(alarmRecyclerViewAdapter.removeItem(position));
                        return false;
                    }
                });
    }

}

File: app/src/main/java/com/example/solar_alarm/AlarmList/AlarmListFragment.kt
package com.example.solar_alarm.AlarmList

import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.test.core.app.ApplicationProvider
import com.example.solar_alarm.Data.ViewModels.*
import com.example.solar_alarm.R
import com.example.solar_alarm.Service.GpsTracker
import com.example.solar_alarm.SolarAlarmApp
import java.time.ZoneId
import java.util.*

@RequiresApi(Build.VERSION_CODES.O)
class AlarmListFragment : Fragment(), OnToggleAlarmListener {

    private val locationViewModel: LocationViewModel by viewModels {
        LocationViewModelFactory((ApplicationProvider.getApplicationContext() as SolarAlarmApp).locationRepository)
    }

    private val solarTimeViewModel: SolarTimeViewModel by viewModels {
        SolarTimeViewModelFactory((ApplicationProvider.getApplicationContext() as SolarAlarmApp).solarTimeRepository)
    }
    // https://github.com/android/compose-samples
    private val solarAlarmViewModel: SolarAlarmViewModel by viewModels {
        SolarAlarmViewModelFactory((ApplicationProvider.getApplicationContext() as SolarAlarmApp).solarAlarmRepository)
    }

    private var alarmRecyclerViewAdapter: AlarmRecycleViewAdapter? = null
    private lateinit var alarmsRecyclerView: RecyclerView
    private var gpsTracker: GpsTracker? = null
    var latitude: TextView? = null
    var longitude: TextView? = null
    private var zoneId: ZoneId? = null
    @RequiresApi(api = Build.VERSION_CODES.O)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        alarmRecyclerViewAdapter = AlarmRecycleViewAdapter(this)
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_listalarms, container, false)
        alarmsRecyclerView = view.findViewById(R.id.fragment_listalarms_recylerView)
        alarmsRecyclerView.setLayoutManager(LinearLayoutManager(context))
        alarmsRecyclerView.setAdapter(alarmRecyclerViewAdapter)
        zoneId = TimeZone.getDefault().toZoneId()
        latitude = view.findViewById(R.id.fragment_listalarms_latitude)
        longitude = view.findViewById(R.id.fragment_listalarms_longitude)

        getLocation(view)

        return view
    }

//    override fun onToggle(alarm: Alarm) {
//        if (alarm.isStarted) {
//            alarm.cancelAlarm(context)
//            //alarmsListViewModel.update(alarm);
//        } else {
//            alarm.schedule(context)
//            //alarmsListViewModel.update(alarm);
//        }
//    }

    fun getLocation(view: View) {
        gpsTracker = GpsTracker(view.context)
        if (gpsTracker!!.canGetLocation()) {
            val lat = gpsTracker!!.latitude
            val lon = gpsTracker!!.longitude
            latitude!!.text = lat.toString()
            longitude!!.text = lon.toString()
        } else {
            gpsTracker!!.showSettingsAlert()
        }
    }
/*
    private fun configureOnClickRecyclerView() {
        ItemClickSupport.addTo(alarmsRecyclerView, R.layout.item_alarm)
                .setOnItemClickListener(ItemClickSupport.OnItemClickListener { recyclerView, position, v ->
                    val alarm = alarmRecyclerViewAdapter!!.getAlarm(position)
                    val bundle = Bundle()
                    bundle.putInt("position", position)
                    val updateAlarmFragment = UpdateAlarmFragment()
                    updateAlarmFragment.arguments = bundle
                    val manager = fragmentManager
                    manager!!.beginTransaction().replace(R.id.activity_main_nav_host_fragment, updateAlarmFragment).commit()
                })
        ItemClickSupport.addTo(alarmsRecyclerView, R.layout.item_alarm)
                .setOnItemLongClickListener(ItemClickSupport.OnItemLongClickListener { recyclerView, position, v ->
                    val alarm = alarmRecyclerViewAdapter!!.getAlarm(position)
                    // 2 - Show result in a Toast
                    //Toast.makeText(getContext(), "You long clicked on user : "+alarm.getTitle(), Toast.LENGTH_SHORT).show();
                    //alarmsListViewModel.delete(alarmRecyclerViewAdapter.removeItem(position));
                    false
                })
    }
    */
}
